juju/juju
# [JUJU-131] Ensure machine removal deletes link-layer device provider IDs

Removing and re-adding a machine based on the same provider instance resulted in the following errors in the controller logs:
```
machine-0: 16:51:48 INFO juju.apiserver.common.networkingcommon processing link-layer devices for machine "1"
machine-0: 16:51:48 WARNING juju.apiserver.instancepoller not setting provider ID for device "ens4" to "17"; it is assigned to another device
machine-0: 16:51:48 WARNING juju.apiserver.instancepoller not setting provider ID for device "ens7" to "18"; it is assigned to another device
```

This was due to the global provider IDs for link-layer devices not being removed with the machine. The reason for this ws that we were only selecting the device document ID when identifying deletion ops, and missing the fact that these had provider IDs.

The bug is corrected here.

In addition we ensure that force-destroying a machine removes its link-layer devices.

Upgrade steps are included to ensure:
- Orphaned link-layer devices are removed.
- Orphaned link-layer device provider IDs are removed.

## QA steps

- Bootstrap to MAAS.
- Add a machine.
- Remove the machine.
- Add the machine again, ensuring the same MAAS instance.
- Observe the absence of WARNING logs above.
- Remove the machine with `--force`.
- Connect to Mongo and check that its link-layer devices are removed with it.
- Add the machine again.
- Destroy the model with force and check Mongo again to see that its devices are removed.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1898195
